File: testers/i1366_position_check.cpp
// #1366: where the dart landed, held at the seam that writes it.
//
// `TurnausClient::detectionBody` is the one place a detection body is composed, and #822
// rule 3 means those bytes are posted FOR EVER -- the spool file keeps them and every
// retry after every restart sends them again. So the whole of what a body says has to be
// decided there, and this file is what holds it: no client, no socket, no spool.
//
// The needles are proved before their absence means anything (#708's shape). A MISS is
// shown to drop a position that a scoring dart WITH THE SAME NUMBERS is shown to carry,
// so "no board_radius in a miss" is a refusal rather than a body that never had one. The
// float just below 360 is shown to be ABOVE the door's published bound before it is shown
// to leave as 0, so the wrap is closing a real 422 rather than tidying.
//
// #1365's rules are quoted here as the server really spells them --
// `App\Autoscoring\PushedPosition::rules()` and `Domain\Scoring\Darts\BoardPosition` --
// so a server that moves a bound makes this file's control wrong rather than silently
// stale, which is the arrangement #1347 made for the sector grammar.
//
//   g++ -std=c++17 -I src -I src/utils -I build/_deps/nlohmann_json-src/include \
//       -I build/_deps/httplib-src testers/i1366_position_check.cpp \
//       src/communication/turnaus_client.cpp -o position_check \
//       $(pkg-config --cflags --libs opencv4) -lpthread

#include <cmath>
#include <iostream>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include "communication/turnaus_client.hpp"

using json = nlohmann::json;

static int checks = 0;
static std::vector<std::string> failures;

static bool say(bool ok, const std::string &what)
{
    checks++;
    std::cout << (ok ? "PASS " : "FAIL ") << what << std::endl;
    if (!ok)
    {
        failures.push_back(what);
    }
    return ok;
}

/** A detection the way the geometry hands one over, with nothing about the board known. */
static DetectorResult dart(const std::string &score)
{
    DetectorResult r;
    r.dart_detected = true;
    r.score = score;
    r.confidence = 0.9f;
    r.camera_index = 1;
    return r;
}

static DetectorResult placed(const std::string &score, float radius, float angle)
{
    DetectorResult r = dart(score);
    r.board_radius_known = true;
    r.board_angle_known = true;
    r.board_radius = radius;
    r.board_angle = angle;
    return r;
}

// ---- #1365's door, reimplemented from its own source ------------------------------------
//
// `PushedPosition::rules()`:
//   board_radius: required_with:board_angle, nullable, numeric, min:0.0,  max:1.2
//   board_angle:  required_with:board_radius, nullable, numeric, min:0.0, max:359.9999
//
// Laravel's `required_with:other` is satisfied only when this field is present AND not
// null, and it is skipped only when `other` is itself absent or null. So an explicit pair
// of nulls is no position, and ONE null beside one number is half a position and a 422.
// A 422 is the expensive refusal here: `TurnausClient::deliver` drops a 422 rather than
// retrying it, so a body this function refuses is a DART lost, not a position.
struct DoorVerdict
{
    bool accepted = true;
    std::string because;
};

static DoorVerdict theDoor(const std::string &body)
{
    DoorVerdict out;
    json j;
    try
    {
        j = json::parse(body);
    }
    catch (const std::exception &e)
    {
        out.accepted = false;
        out.because = std::string("the body is not JSON: ") + e.what();
        return out;
    }
    const bool has_radius = j.contains("board_radius") && !j["board_radius"].is_null();
    const bool has_angle = j.contains("board_angle") && !j["board_angle"].is_null();
    if (has_radius != has_angle)
    {
        out.accepted = false;
        out.because = "required_with: half a polar position";
        return out;
    }
    if (!has_radius)
    {
        return out; // absent, or an explicit pair of nulls: no position, and admitted
    }
    if (!j["board_radius"].is_number() || !j["board_angle"].is_number())
    {
        out.accepted = false;
        out.because = "numeric";
        return out;
    }
    const double radius = j["board_radius"].get<double>();
    const double angle = j["board_angle"].get<double>();
    if (!(radius >= 0.0) || !(radius <= 1.2))
    {
        out.accepted = false;
        out.because = "board_radius is outside [0, 1.2]: " + body;
        return out;
    }
    if (!(angle >= 0.0) || !(angle <= 359.9999))
    {
        out.accepted = false;
        out.because = "board_angle is outside [0, 359.9999]: " + body;
        return out;
    }
    return out;
}

static json bodyOf(const DetectorResult &r)
{
    return json::parse(TurnausClient::detectionBody("01HZZZZZZZZZZZZZZZZZZZZZZZ", r).json);
}

int main()
{
    const std::string REF = "01HZZZZZZZZZZZZZZZZZZZZZZZ";

    // ---- what a body has always said, unchanged --------------------------------------
    {
        TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, dart("T20"));
        json j = json::parse(b.json);
        say(j["reference"] == REF, "the reference passed in is the reference sent");
        say(j["sector"] == "T20", "the sector is #821's grammar");
        say(j["bounced_out"] == false, "bounced_out is still there and still false");
        say(!b.carries_position && b.position_withheld.empty(),
            "a detection that knows no position carries none and withholds nothing");
        say(b.json.find("board_radius") == std::string::npos &&
                b.json.find("board_angle") == std::string::npos,
            "and NEITHER KEY is in the bytes -- absent, never a nought");
    }

    // ---- an unpostable sector is still no body at all ---------------------------------
    for (const std::string &score : {"", "S21", "X5"})
    {
        say(TurnausClient::detectionBody(REF, placed(score, 0.5f, 90.0f)).json.empty(),
            "'" + score + "' still answers with no body, position or no position");
    }

    // ---- THE CONTROL: a scoring dart carries both keys, and the numbers are the ones --
    {
        TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, placed("T20", 0.6123f, 12.3456f));
        json j = json::parse(b.json);
        say(b.carries_position, "a dart that knows where it landed says so");
        say(j.contains("board_radius") && j.contains("board_angle"), "and both keys are in the body");
        say(std::fabs(j["board_radius"].get<double>() - 0.6123) < 1e-9 &&
                std::fabs(j["board_angle"].get<double>() - 12.3456) < 1e-9,
            "with the radius and the angle the geometry measured");
        say(b.json.find("\"board_radius\":0.6123") != std::string::npos &&
                b.json.find("\"board_angle\":12.3456") != std::string::npos,
            "written at the four places the columns keep, not as the float's binary truth "
            "(" + b.json + ")");
        say(b.position_withheld.empty(), "and nothing is withheld");
    }

    // ---- the float's binary truth, which is what makes the rounding load-bearing ------
    {
        json raw;
        raw["board_radius"] = (double)0.1f;
        say(raw.dump().find("0.10000000149011612") != std::string::npos,
            "the needle: a float promoted straight into a body prints its binary truth "
            "(" + raw.dump() + ")");
        say(bodyOf(placed("S1", 0.1f, 0.1f)).dump().find("0.1,") != std::string::npos ||
                bodyOf(placed("S1", 0.1f, 0.1f))["board_radius"].dump() == "0.1",
            "and the seam prints 0.1");
    }

    // ---- BOTH OR NEITHER, which is the refusal that costs a dart ----------------------
    {
        DetectorResult radius_only = dart("Bull");
        radius_only.board_radius_known = true;
        radius_only.board_radius = 0.0212f;
        TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, radius_only);
        say(!b.carries_position && b.json.find("board_") == std::string::npos,
            "a bull that knows its radius and has no measured wedge sends NEITHER key");
        say(b.position_withheld.find("radius") != std::string::npos && !b.position_out_of_bounds,
            "and says which half it had, at debug rather than as a fault (" + b.position_withheld + ")");

        DetectorResult angle_only = dart("S20");
        angle_only.board_angle_known = true;
        angle_only.board_angle = 3.5f;
        TurnausClient::DetectionBody c = TurnausClient::detectionBody(REF, angle_only);
        say(!c.carries_position && c.json.find("board_") == std::string::npos,
            "and an angle without a radius is the same refusal the other way round");
        say(c.position_withheld.find("angle") != std::string::npos,
            "named the other way round too (" + c.position_withheld + ")");
    }

    // ---- A MISS HAS NO PLACE, and the needle is proved before its absence means anything
    {
        DetectorResult miss = placed("MISS", 0.6123f, 12.3456f);
        TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, miss);
        json j = json::parse(b.json);
        say(j["sector"] == "None", "a MISS is still #821's None");

        // The needle exists: the SAME numbers on a scoring dart really do reach a body.
        DetectorResult scoring = placed("T20", 0.6123f, 12.3456f);
        say(TurnausClient::detectionBody(REF, scoring).carries_position,
            "the needle: those same numbers on a scoring dart DO reach the body");
        say(!b.carries_position, "so the miss dropping them is a refusal and not an empty result");
        say(b.json.find("0.6123") == std::string::npos && b.json.find("12.3456") == std::string::npos,
            "swept BY VALUE as well as by key: neither number is anywhere in the miss's bytes "
            "(" + b.json + ")");
        say(!b.position_withheld.empty() && b.position_withheld.find("miss") != std::string::npos,
            "and the withholding says so (" + b.position_withheld + ")");

        // None is the same sector by another name, and a board may already send it.
        say(!TurnausClient::detectionBody(REF, placed("None", 0.5f, 5.0f)).carries_position,
            "and a detection already spelled None is the same refusal");
    }

    // ---- 360 IS 0, which is the one value this footage could ever argue about ---------
    {
        // THE WHOLE INTERVAL, counted rather than sampled. `board_angle` is a float, and
        // (359.9999, 360) holds exactly THREE of them. Every one prints, through nlohmann,
        // as a decimal above the door's published bound -- so every one is a 422, and a
        // 422 is a dart. Rounded to the four places the column really keeps, ONE of them
        // (360 - 1 ulp, which is 359.99995 or more) becomes 360.0000 and leaves as 0; the
        // other two become 359.9999, which is a place. Both answers are admitted, and it
        // is worth knowing which is which rather than asserting the same thing three times.
        int refused_raw = 0;
        int sent_as_zero = 0;
        int sent_as_the_bound = 0;
        float f = std::nextafter(360.0f, 0.0f);
        for (int i = 0; i < 3; i++)
        {
            say((double)f > 359.9999 && (double)f < 360.0,
                "the float " + std::to_string(i + 1) + " below 360 really is in (359.9999, 360)");

            json raw;
            raw["board_angle"] = (double)f;
            DoorVerdict unrounded = theDoor("{\"board_radius\":0.5,\"board_angle\":" +
                                            raw["board_angle"].dump() + "}");
            if (!unrounded.accepted)
            {
                refused_raw++;
            }

            TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, placed("S20", 0.5f, f));
            json j = json::parse(b.json);
            say(b.carries_position && theDoor(b.json).accepted,
                "and the seam sends it as something the door admits (" + j["board_angle"].dump() + ")");
            if (j["board_angle"].dump() == "0.0")
            {
                sent_as_zero++;
            }
            else if (j["board_angle"].dump() == "359.9999")
            {
                sent_as_the_bound++;
            }
            f = std::nextafter(f, 0.0f);
        }
        say(refused_raw == 3,
            "the needle: ALL THREE floats between 359.9999 and 360 are refused by the door as "
            "published, unrounded -- the 422 this wrap closes is real (" +
                std::to_string(refused_raw) + " of 3)");
        say(sent_as_zero == 1,
            "exactly one of them rounds into 360.0000 and is sent as 0, because 360 degrees is "
            "0 degrees (" + std::to_string(sent_as_zero) + ")");
        say(sent_as_the_bound == 2,
            "and the other two round to 359.9999, which is a place the column holds (" +
                std::to_string(sent_as_the_bound) + ")");
        // A rounded negative nothing is a real double with a sign on it, and it prints
        // with the sign. It would validate; it reads like a fault.
        json tiny = bodyOf(placed("S20", -0.00001f, -0.00001f));
        say(tiny["board_radius"].dump() == "0.0" && tiny["board_angle"].dump() == "0.0",
            "a radius and an angle that round to nothing leave as 0.0 and never as -0.0 ("
                + tiny.dump() + ")");
    }

    // ---- the bounds themselves, at the places they are really kept --------------------
    {
        say(bodyOf(placed("S20", 1.2f, 359.9999f)).contains("board_radius"),
            "1.2 is a place and 359.9999 is a place: the published bounds are inclusive");
        say(bodyOf(placed("S20", 0.0f, 0.0f))["board_radius"].dump() == "0.0",
            "the bull centre is radius 0 and it is sent as 0, never as an absence");

        TurnausClient::DetectionBody over = TurnausClient::detectionBody(REF, placed("S20", 1.25f, 10.0f));
        say(!over.carries_position && over.position_out_of_bounds,
            "1.25 is off the far end, so the POSITION is withheld");
        say(over.json.find("\"sector\":\"S20\"") != std::string::npos,
            "and THE DART STILL GOES -- a 422 is not retried, so a refused body is a lost dart");
        say(over.position_withheld.find("1.2500") != std::string::npos,
            "the refusal names the number it refused (" + over.position_withheld + ")");

        // 1.20004 rounds INTO the bound at the precision the column keeps, so it is a
        // place; 1.20006 does not.
        say(bodyOf(placed("S20", 1.20004f, 10.0f)).contains("board_radius"),
            "1.20004 is 1.2000 at four places, which is a place the column holds");
        say(!TurnausClient::detectionBody(REF, placed("S20", 1.20006f, 10.0f)).carries_position,
            "1.20006 is 1.2001, which is not");
    }

    // ---- POSTED FOR EVER: the bytes are settled here and nowhere else -----------------
    {
        DetectorResult r = placed("D18", 0.9876f, 201.5f);
        std::string once = TurnausClient::detectionBody(REF, r).json;
        std::string twice = TurnausClient::detectionBody(REF, r).json;
        say(once == twice, "the same detection and the same reference build the same bytes");
        say(json::parse(once).contains("board_radius"),
            "and those bytes -- the ones the spool file keeps and every retry re-posts -- "
            "are the ones the position is in");
    }

    // ---- EVERY body this seam can produce satisfies the door ---------------------------
    // #1347's closing argument: nothing it lets out can meet a 422, so a dart is never
    // lost between a green detector and a green server.
    {
        int refused = 0;
        std::string first;
        const float radii[] = {-1.0f, 0.0f, 0.00004f, 0.25f, 0.5f, 0.9999f, 1.0f, 1.19996f,
                               1.2f, 1.20004f, 1.20006f, 1.25f, 3.0f};
        const float angles[] = {-1.0f, 0.0f, 0.00004f, 9.0f, 179.99995f, 180.0f, 359.0f,
                                359.99985f, 359.9999f, 359.99995f,
                                std::nextafter(360.0f, 0.0f)};
        for (const std::string &score : {"S1", "T20", "D18", "BULL", "OUTER", "MISS", "25", "Bull", "None"})
        {
            for (float radius : radii)
            {
                for (float angle : angles)
                {
                    for (int known = 0; known < 4; known++)
                    {
                        DetectorResult r = dart(score);
                        r.board_radius_known = (known & 1) != 0;
                        r.board_angle_known = (known & 2) != 0;
                        r.board_radius = radius;
                        r.board_angle = angle;
                        TurnausClient::DetectionBody b = TurnausClient::detectionBody(REF, r);
                        if (b.json.empty())
                        {
                            continue;
                        }
                        DoorVerdict v = theDoor(b.json);
                        if (!v.accepted)
                        {
                            refused++;
                            if (first.empty())
                            {
                                first = b.json + " -- " + v.because;
                            }
                        }
                    }
                }
            }
        }
        say(refused == 0, "every body the seam can build is admitted by #1365's door ("
                              + std::to_string(refused) + " refused" +
                              (first.empty() ? "" : ", first: " + first) + ")");
    }

    std::cout << std::endl
              << "checks " << checks << "   failed " << failures.size() << std::endl;
    for (const std::string &f : failures)
    {
        std::cout << "  - " << f << std::endl;
    }
    return failures.empty() ? 0 : 1;
}

File: testers/i1366_position_check.py
#!/usr/bin/env python3
"""#1366: the dart's place on the board, from the seam that writes it to the wire.

Three cases, and they are three different questions rather than one asked three ways.

  seam    testers/i1366_position_check.cpp: what a body may say, held against #1365's
          rules quoted as the server spells them. Pure -- no client, no socket, no spool.
  spool   THE CRITERION A HAPPY PATH CANNOT PROVE. #822 rule 3 writes a body once, when
          the dart is offered, and posts those same bytes for ever. So a position added
          at POST time is a position every dart thrown while the network was down loses,
          and a run against a reachable server would look perfect. Here three placed
          darts are offered at a CLOSED PORT, the bytes on the disk are read, and a
          SECOND PROCESS over the same spool delivers them to a stub that says what
          arrived.
  rig     mocks/rig-20260918, the real binary, #822's stub: every pushed body held to
          the BOARD log line the same detection printed.

Runs inside od-amd64:bullseye with the worktree at /app and a Linux build in /app/build:
  python3 /app/testers/i1366_position_check.py [case ...]
Exit code 0 only when every check passed. Prints `checks N   failed M`.
"""
import json
import os
import re
import socket
import subprocess
import sys
import time

APP = os.environ.get("OD_APP", "/app")
BIN = os.environ.get("OD_BIN", APP + "/build/opendartboard")
STUB = APP + "/testers/turnaus_stub.py"
WORK = os.environ.get("CHECK_WORK", "/tmp/i1366-check")
RIG = ",".join(APP + "/mocks/rig-20260918/cam_%d.mp4" % i for i in (1, 2, 3))
RIG_CYCLES = int(os.environ.get("OD_RIG_CYCLES", "2000"))

CLUB = "483920"
ANSI = re.compile(r"\x1b\[[0-9;]*m")

# What score_processing prints for every result it publishes, once each. `to_string(float)`
# gives six places; a body carries the four the columns keep, so the comparison rounds.
BOARD_LINE = re.compile(
    r"BOARD: (?P<wedge>wedge measured|wedge by default) \| ring=(?P<ring>[^|]*)\| "
    r"segment=(?P<segment>-?\d+) \| radius=(?P<radius>none|-?[0-9.]+) \| angle=(?P<angle>none|-?[0-9.]+)")

# The four decimal places #1365 keeps, in the one spelling this file uses for both sides.
PLACES = 4

checks = 0
failures = []


def check(ok, name):
    global checks
    checks += 1
    print(("PASS " if ok else "FAIL ") + name, flush=True)
    if not ok:
        failures.append(name)
    return ok


def at_places(value):
    """The four places a body carries, as a float, for comparing with a log line."""
    return round(float(value), PLACES)


class Stub:
    def __init__(self, case, port, **env):
        self.dir = os.path.join(WORK, case)
        os.makedirs(self.dir, exist_ok=True)
        self.transcript = os.path.join(self.dir, "transcript.jsonl")
        e = dict(os.environ)
        e.update({"STUB_PORT": str(port), "STUB_TRANSCRIPT": self.transcript,
                  "STUB_SAMPLE_SECONDS": "0", "STUB_INTERVAL_SECONDS": "5",
                  "STUB_SILENCE_SECONDS": "60"})
        e.update({k: str(v) for k, v in env.items()})
        self.err = open(os.path.join(self.dir, "stub.err"), "w")
        self.proc = subprocess.Popen([sys.executable, STUB], env=e, stderr=self.err, stdout=self.err)
        deadline = time.time() + 10
        while time.time() < deadline:
            try:
                socket.create_connection(("127.0.0.1", port), timeout=0.2).close()
                return
            except OSError:
                time.sleep(0.1)
        raise RuntimeError("stub did not listen")

    def events(self, name=None):
        out = []
        try:
            with open(self.transcript) as fh:
                for line in fh:
                    if line.strip():
                        ev = json.loads(line)
                        if name is None or ev["event"] == name:
                            out.append(ev)
        except FileNotFoundError:
            pass
        return out

    def stop(self):
        self.proc.terminate()  # its own pid, nothing else
        self.proc.wait(timeout=10)
        self.err.close()


class Board:
    """One board's HOME: its credential, its spool, and every process run against them."""

    def __init__(self, case, port):
        self.case = case
        self.port = port
        self.dir = os.path.join(WORK, case)
        self.cwd = os.path.join(self.dir, "cwd")
        os.makedirs(self.cwd, exist_ok=True)
        self.config = os.path.join(self.dir, "config", "opendartboard")
        os.makedirs(self.config, exist_ok=True)
        self.credentials = os.path.join(self.config, "credentials.json")
        self.spool = os.path.join(self.config, "owed.jsonl")
        self.env = {"PATH": os.environ.get("PATH", "/usr/bin:/bin"), "HOME": self.dir,
                    "XDG_CONFIG_HOME": os.path.join(self.dir, "config")}
        self.text = ""

    def pair(self, flag, code, port=None):
        out = os.path.join(self.dir, "pair.out")
        with open(out, "w") as fh:
            return subprocess.run([BIN, flag, code, "--turnaus",
                                   "http://127.0.0.1:%d" % (port or self.port), "--allow-plaintext"],
                                  cwd=self.cwd, env=self.env, stdin=subprocess.DEVNULL,
                                  stdout=fh, stderr=subprocess.STDOUT, timeout=60).returncode

    def run(self, cycles, cams, label="run"):
        env = dict(self.env)
        env["OD_MAX_CYCLES"] = str(cycles)
        out = os.path.join(self.dir, label + ".out")
        err = os.path.join(self.dir, label + ".err")
        with open(out, "w") as o, open(err, "w") as e:
            rc = subprocess.run([BIN, "--debug", "--cams", cams, "--width", "1280", "--height", "720",
                                 "--allow-plaintext"],
                                cwd=self.cwd, env=env, stdin=subprocess.DEVNULL,
                                stdout=o, stderr=e, timeout=1800).returncode
        body = ""
        for path in (out, err):
            with open(path) as fh:
                body += ANSI.sub("", fh.read())
        self.text = body
        return rc

    def harness(self, binary, mode, port=None, label=None):
        """Run the #1366 client harness over THIS board's config dir."""
        out = os.path.join(self.dir, (label or mode) + ".out")
        with open(out, "w") as fh:
            rc = subprocess.run([binary, "http://127.0.0.1:%d" % (port or self.port),
                                 self.credentials, mode],
                                cwd=self.cwd, env=self.env, stdin=subprocess.DEVNULL,
                                stdout=fh, stderr=subprocess.STDOUT, timeout=300).returncode
        with open(out) as fh:
            return rc, ANSI.sub("", fh.read())

    def spool_records(self):
        out = []
        try:
            with open(self.spool) as fh:
                for line in fh:
                    if line.strip():
                        out.append(json.loads(line))
        except FileNotFoundError:
            pass
        return out


def compile_harness(case, source, binary_name):
    """Build one of #1366's harnesses against THIS tree's own client."""
    binary = os.path.join(WORK, case, binary_name)
    os.makedirs(os.path.dirname(binary), exist_ok=True)
    line = (
        "g++ -std=c++17 -I {app}/src -I {app}/src/utils "
        "-I {app}/build/_deps/nlohmann_json-src/include -I {app}/build/_deps/httplib-src "
        "$(pkg-config --cflags opencv4 2>/dev/null) "
        "{app}/testers/{src} {app}/src/communication/turnaus_client.cpp -o {bin} "
        "$(pkg-config --libs opencv4 2>/dev/null) -lpthread"
    ).format(app=APP, src=source, bin=binary)
    built = subprocess.run(line, shell=True, capture_output=True, text=True, timeout=900)
    if not check(built.returncode == 0, "%s: %s compiles (%s)"
                 % (case, source, built.stderr.strip().splitlines()[-1:] or "no output")):
        return None
    return binary


def position_in(body):
    """The pair a body carries, or None. Half a pair is neither -- it is a 422."""
    has_radius = "board_radius" in body and body["board_radius"] is not None
    has_angle = "board_angle" in body and body["board_angle"] is not None
    if has_radius != has_angle:
        return "half"
    if not has_radius:
        return None
    return (at_places(body["board_radius"]), at_places(body["board_angle"]))


def case_seam():
    """The pure seam, which is where every byte of a body is decided."""
    binary = compile_harness("seam", "i1366_position_check.cpp", "position_check")
    if binary is None:
        return
    run = subprocess.run([binary], capture_output=True, text=True, timeout=300)
    print(run.stdout, flush=True)
    tail = [l for l in run.stdout.splitlines() if l.startswith("checks ")]
    check(run.returncode == 0, "seam: every check at the body-building seam passed (%s)"
          % (tail[-1] if tail else "no summary at all"))


def case_spool():
    """A dart delivered after a restart keeps its position.

    The body is written ONCE, at the offer, and posted for ever. So this is asked of the
    bytes on the disk -- written by a process that never reached a server -- and then of
    what a second process delivers out of them. A position put in at POST time would leave
    the spool file clean and still arrive at a reachable stub: the first half of this case
    is the only place that difference is visible.
    """
    port = 18661
    gone = 18670  # nothing has ever listened here
    stub = Stub("spool", port)
    board = Board("spool", port)
    check(board.pair("--pair", CLUB) == 0, "spool: the board pairs to its club")

    # ONE server for both halves, and the board's network goes away rather than the
    # server. Stopping the stub instead would re-pair against a second stub process,
    # which mints a new club token generation and answers the board's own credential
    # with a 401 -- measured, and it is #1259's rule working rather than this case's
    # subject. What is wanted here is a board that cannot reach a Turnaus that is fine.

    binary = compile_harness("spool", "i1366_spool_check.cpp", "spool_check")
    if binary is None:
        return

    rc, text = board.harness(binary, "offer", port=gone)
    check(rc == 0, "spool: three placed darts and a takeout were offered at a closed port (rc=%s)" % rc)
    check("delivered=0" in text, "spool: and nothing was delivered, because nothing was listening")

    # ---- THE ASSERTION, made against the disk ----------------------------------------
    records = board.spool_records()
    detections = [r for r in records if r["path"].endswith("/detections")]
    check(len(detections) == 3, "spool: three detections are on the disk (%d records, %d detections)"
          % (len(records), len(detections)))

    # The three the harness offers, named here so the sweep is by VALUE and not only by key.
    wanted = {"T20": (0.6123, 12.3456), "D18": (0.9876, 201.5), "Bull": (0.0212, 44.25)}
    on_disk = {}
    for record in detections:
        body = json.loads(record["body"])
        on_disk[body.get("sector")] = position_in(body)
    check(on_disk == wanted,
          "spool: EVERY position is in the bytes on the disk, by key and by value -- a position "
          "added at POST time would leave this file clean (%s)" % on_disk)

    takeouts = [r for r in records if r["path"].endswith("/takeouts")]
    check(len(takeouts) == 1 and takeouts[0]["body"] == "{}",
          "spool: and a takeout body is still exactly {} (%s)" % [t["body"] for t in takeouts])

    # ---- and then of what a SECOND PROCESS delivers out of them ------------------------
    rc, text = board.harness(binary, "resume", label="resume")
    check(rc == 0, "spool: a second process over the same spool delivers the round (rc=%s)" % rc)

    counted = stub.events("counted")
    delivered = {}
    for event in counted:
        delivered[event["sector"]] = position_in(event["body"])
    check(delivered == wanted,
          "spool: and what ARRIVED after the restart carries every position, by key and by value (%s)"
          % delivered)
    check(len(stub.events("takeout")) == 1,
          "spool: the takeout that closes the round arrives too, so this was a whole round")
    stub.stop()


def case_rig():
    """mocks/rig-20260918, the real binary, and the BOARD lines held to what was pushed.

    One BOARD line is printed for every result score_processing publishes, and every one
    of those is offered, so the two sequences are the same sequence read at two ends of
    the program. What is asked of each pair is #1366's whole contract: a known position
    reaches the body at four places, an unknown one reaches it as NO KEY AT ALL rather
    than as a nought, a MISS carries none whatever it knew, and half a position is never
    sent.
    """
    port = 18662
    stub = Stub("rig", port)
    board = Board("rig", port)
    check(board.pair("--pair", CLUB) == 0, "rig: the board pairs to its club")
    rc = board.run(RIG_CYCLES, RIG)
    check(rc == 0, "rig: the run ends normally (rc=%s)" % rc)

    lines = [m.groupdict() for m in BOARD_LINE.finditer(board.text)]
    counted = stub.events("counted")
    print("rig: %d BOARD lines, %d pushed detections" % (len(lines), len(counted)), flush=True)
    for line in lines:
        print("   BOARD ring=%(ring)s segment=%(segment)s radius=%(radius)s angle=%(angle)s" % line,
              flush=True)
    for event in counted:
        print("   PUSH  " + json.dumps(event["body"], sort_keys=True), flush=True)

    if not check(len(lines) > 0 and len(counted) > 0,
                 "rig: the footage scored at least one dart and at least one was pushed "
                 "(%d BOARD lines, %d pushes)" % (len(lines), len(counted))):
        stub.stop()
        return
    if not check(len(lines) == len(counted),
                 "rig: one BOARD line per pushed detection -- the two ends of the same sequence "
                 "(%d and %d)" % (len(lines), len(counted))):
        stub.stop()
        return

    matched = 0
    absent = 0
    for line, event in zip(lines, counted):
        body = event["body"]
        got = position_in(body)
        a_miss = event["sector"] == "None"
        knows_both = line["radius"] != "none" and line["angle"] != "none"
        if knows_both and not a_miss:
            expected = (at_places(line["radius"]), at_places(line["angle"]))
            check(got == expected,
                  "rig: %s was pushed at the place its BOARD line printed (%s vs %s)"
                  % (event["sector"], got, expected))
            matched += 1
        else:
            check(got is None,
                  "rig: %s knew %s, so it was pushed with NEITHER key rather than a nought (%s)"
                  % (event["sector"], "nothing" if not knows_both else "a place it may not send", got))
            absent += 1
    check(all(position_in(e["body"]) != "half" for e in counted),
          "rig: nothing was pushed with half a position, which the door refuses with 422")
    check(all(position_in(e["body"]) is None for e in counted if e["sector"] == "None"),
          "rig: every MISS was pushed without a place on the board")
    print("rig: %d pushes carried a position, %d carried none" % (matched, absent), flush=True)
    stub.stop()


CASES = {"seam": case_seam, "spool": case_spool, "rig": case_rig}

if __name__ == "__main__":
    os.makedirs(WORK, exist_ok=True)
    wanted = sys.argv[1:] or list(CASES)
    for name in wanted:
        if name not in CASES:
            print("no such case: %s; the cases are %s" % (name, ", ".join(CASES)))
            sys.exit(2)
    for name in wanted:
        print("\n==== %s ====" % name, flush=True)
        CASES[name]()
    print("\nchecks %d   failed %d" % (checks, len(failures)))
    for f in failures:
        print("  - " + f)
    sys.exit(1 if failures else 0)
